main.go

```diff
@@ -0,0 +1,209 @@
+	rowOffset   int = 1
+	sheetOffset int = 1
+)
+
+type Input struct {
+	Filename       string
+	SheetName      string
+	UsernameHeader string
+	PasswordHeader string
+	SheetPassword  string
+}
+
+type Portal struct {
+	*Input
+	Environment string
+	Hostname    string
+	IdmHostname string // identity management hostname
+	infoLog     *log.Logger
+	errorLog    *log.Logger
+}
+
+type Creds struct {
+	OldPassword string
+	NewPassword string
+}
+
+var client *http.Client
+
+func init() {
+	jar, err := cookiejar.New(nil)
+	if err != nil {
+		log.Fatal("error creating cookiejar")
+	}
+	client = &http.Client{
+		Jar: jar,
+	}
+}
+
+func resetPasswords(f *excelize.File, config *Portal) (err error) {
+	defer func() error {
+		if rerr := recover(); rerr != nil && fmt.Sprint(rerr) == ErrCryptoSourceFailure.Error() {
+			return ErrCryptoSourceFailure
+		} else if rerr != nil {
+			panic(rerr)
+		}
+		return err
+	}()
+
+	rows, err := f.GetRows(automatedSheet)
+	if err != nil {
+		return err
+	}
+
+	var now time.Time
+
+	numSuccess := 0
+	numFail := 0
+	numNoRotation := 0
+
+	var lastRotated time.Time
+
+	randomPasswords := make([]string, len(rows)-rowOffset)
+	for i := 0; i < len(rows)-rowOffset; i++ {
+		password := getRandomPassword()
+		randomPasswords[i] = password
+	}
+
+	for i, row := range rows[rowOffset:] {
+		now = time.Now().UTC()
+		name := row[user]
+
+		if row[timestamp] == "Rotate Now" {
+			// force rotation
+			lastRotated = now.AddDate(0, 0, -thirtyDays-2)
+		} else {
+			lastRotated, err = time.Parse(time.UnixDate, row[timestamp])
+			if err != nil {
+				config.errorLog.Printf("error parsing timestamp from row %d for user %s: %v", i+rowOffset, name, err)
+				return err
+			}
+		}
+
+		if now.Before(lastRotated.AddDate(0, 0, thirtyDays)) {
+			config.infoLog.Printf("%s: no rotation needed", row[user])
+			numNoRotation++
+			continue
+		} else {
+			newPassword := randomPasswords[i]
+			err = changeUserPassword(client, config, name, row[portal], newPassword)
+			if err != nil {
+				numFail++
+				config.errorLog.Printf("user %s password reset FAIL: %v", name, err)
+				continue
+			}
+			numSuccess++
+			// copy portal col password to previous col
+			err = copyCell(f, config.Filename, portal+sheetOffset, i+rowOffset+sheetOffset, previous+sheetOffset, i+rowOffset+sheetOffset)
+			if err != nil {
+				config.errorLog.Printf("failed to write previous password %s to sheet %s, row %d for user %s: %v",
+					row[portal], config.SheetName, i+rowOffset, name, err)
+				return err
+			}
+
+			// Write new password to portal col
+			err = writeCell(f, config.Filename, automatedSheet, portal+sheetOffset, i+rowOffset+sheetOffset, newPassword)
+			if err != nil {
+				config.errorLog.Printf("failed to write new password to sheet %s in row %d for user %s: %v; manually set password for user",
+					config.SheetName, i+rowOffset+sheetOffset, name, err)
+				return nil
+			}
+			// set timestamp
+			ts := time.Now().UTC().Format(time.UnixDate)
```

Comment:
nit: use `now` instead?

Reply:
The time difference between when the loop starts and when the timestamp is written is on the order of 5 seconds. Since that level of precision doesn't matter, I changed this line to use now.